README
mheg2xmltv
Copyright 2010-2017 by Solorvox <solorvox@epic.geek.nz>
Latest version: https://github.com/solorvox/mheg2xmltv

-----------------------------------------------------------------
Contents:

1.0 - About
1.1 - Redbutton-download
1.2 - mheg2xmltv Setup
1.3 - Listing metadata / categories
1.4 - Additional Tools

-----------------------------------------------------------------

-------------
1.0 - About

A utility to convert MHEG-5 data to an XMLTV file to importing into mythtv for
extended program guide listings.

-------------
1.1 - Redbutton-download

You need a copy of redbutton-download, not browser, that can be found
at http://sourceforge.net/projects/redbutton/files/

After compiling, create your channels.conf.  Follow the documentation provided
by rb-download.  

Just as a tip, you'll need a channels.conf file made using dvbscan (or just scan 
depending on distro) using a base file.  For Christchurch I used:

scan chch > ~/.tzap/channels.conf

-- Cut chch --
# Multiplex DA
T 562000000 8MHz 3/4 NONE QAM64 8k 1/16 NONE
# Multiplex DB
T 578000000 8MHz 3/4 NONE QAM64 8k 1/16 NONE
# Multiplex DC
T 594000000 8MHz 3/4 NONE QAM64 8k 1/16 NONE
-- Cut --

Find your correct values here: 

http://en.wikipedia.org/wiki/Freeview_%28New_Zealand%29#Terrestrial_transmitters

By default, the service ID for New Zealand is 1200 for Channel TVONE.  If you need
to tune your adpater, try using ivtv-tune part of ivtv-utils package on Debian.

ivtv-tune --freqtable=newzealand --channel=1 --device=/dev/video1
or 
ivtv-tune --frequency=45.250 --device=/dev/video1 

Included in the mheg2xmltv/tools directory is a helper script, mheg5grab.sh.
It will assist in downloading MHEG5 data by only running rb-download for a few seconds.
It should be run once per day from cron.  Check the configuration in the script and change
for you setup.  You will need to add an entry to cron.  For example, to run at 02:07AM every day:

07 02 * * * /home/user/mheg2xmltv/tools/mheg5grab.sh

-------------
1.2 - mheg2xmltv Setup

You can either directly edit mheg2xmltv.sh and change the default values, or
pass them via command line arguments.  It is important to set your source directory
where rb-download downloaded the MHEG data.  Additonal, you may override default values on
the command line.  For example, the default channel mapping is included inside the main file.
However, you may pass a map file as "mheg2xmltv -c /path/map" for your own custom mapping.  Follow
the example format included in the mheg2xmltv.sh file.

-------------
1.3 - Listing metadata / categories

Version 0.3.5 adds a new feature to "guess" the show categories by looking at keywords in title and descriptions. 
To enable this feature, you must add -g to the command line.

The order of which categories are assigned are as follows:

1) MHEG2 catgegory data found
2) series-metadata.txt match found via series ID number (use tools/mhegshowid to find ID numbers)
3) series-metadata.txt match found via title match (exact match only so "CIS" won't match "CSI: New York")
4) If "guess" mode was enabled with "-g", keyword matching is done with title and description

If at any point a match is found, processing stops.  So if you do not agree/like the -g mode, you may over-ride
the data by placing the desired information in series-metadata.txt.

-------------
1.4 - Additional Tools

Basic tools have been included in the tools/ subdirectory.  They are not a required for mheg2xmltv,
but are provided as additional utilities.  The $BASE directory variable should be changed to match
the source directory.

mhegshowid - Returns the CRID show ID numbers found
	examples:
	$ mhegshowid "House"
	30774
	$ mhegshowid "The Pacific"
	1006745

mhegsearch - Finds strings matching anywhere in the listing
	example:	
	$ mhegshowid fire
	7|28800|32400|1|08:00|49|/T254983|Magnum, P.I.|'Solo Flight'. Fired from a case...<cut>
	(Matched "Fired" in the description.)
			
mhegview   - Converts MHEG data to human readable format
	example:
	$ mhegview /carousels/1002/1/dir-31-3030366f
	24|77400|81000|1|21:30|49|/SNZ10189342|Sea Patrol|It's a day the crew of ...<cut> 

	* This script is needed by others.

mhegtitle  - Matches only titles that start with string
	example:
	$ mhegtitle House
	28|73800|77400|1|20:30|49|/T284056|House|A high-school senior blacks out ...<cut>

	Note: This only matches titles that start with the string. 	
	* This script is required by others.

mheg5grab.sh - Wrapper/helper script to assist in downloading MHEG5 data using rb-download.
	See setup notes above for details.  Does not take any arguments or return any output
	on normal operation.


Changelog
0 5.0 Updated expected directory to match new format from epg/data to epgdtt/data (20171130)

0.4.0 Changed default method for channel ID to use "ServiceID" instead of crid://<uri> because TVNZ duplicated crid for more than one channel.  Added service ID (eg 1200 for TV One) to the channel data.  Improved feedback from tool scripts.  Improved listing category guess detection and more series metadata added.  Fixed tools for better recursive search of data directory. (20161031)

0.3.6 Added new channel mappings, removed TV7, updated install docs. (20120724)

0.3.5 Fixed another bug with multi-day listings to merge shows, added new logic to "guess" at categories of show based on keywords in title and description fields, added lots of new shows to series-categories.txt (20110525)

0.3.4 Fixed tracker bug 3153962 with greedy sed channel remap. (20110525)

0.3.3 Updated new stations Trackside, Stratos, U, and FOUR, removed channels that no longer exist (TV6 and FreeviewHD Demo) (20110525)

0.3.2 Thanks to David Moore <dmoo1790@ihug.co.nz> for code fixing programs running across midnight, and bug with tool script to kill rb-download child process.  Added new command line option, -c,  for external custom channel map file.  (20100711)
 
0.3.1 Added wrapper script (in tools/) to assist with MHEG5 data downloading.  Updated documentation. (20100607)

0.3.0 Added additional metadata file for show category information.  Added CRID event ID number as xmltv episode-number.  Added utilties in tools/, they are not required for mheg2xmltv.  Added support for dolby audio. (20100501)

0.2.0 Added support for channel id remapping, channel icons, and added channel definition to top of xmltv output. (20100430)

0.1.1 Fixed issue with duplicate CRIDs mapped to same channel id. (20100430)

0.1.0 First public release (20100429)



update-epg.sh
#!/bin/bash

# BASE Directory for mheg5xmltv
DIR="$HOME/mheg2xmltv"

cd $DIR
# First grab the MHEG5 data
./tools/mheg5grab.sh
RESULT=$?

if [ "$RESULT" -ne 0 ]; then
	echo "ERROR: MHEG5 Grab failed with error code: $RESULT"
	exit -1
fi

# Add options such as -g for guess categories here
./mheg2xmltv.sh -g
RESULT=$?

if [ "$RESULT" -ne 0 ]; then
	echo "ERROR: mheg2xmltv conversion failed with error code: $RESULT"
	exit -1
fi

# Update mythtv database
# For myth version 0.24 or less /usr/bin/mythfilldatabase --xmlfile /tmp/xmltv.xml > /dev/null
# Update for 0.25 and later
# For <= mythtv 0.26 use --update instead of  --only-update-guide
/usr/bin/mythfilldatabase  --only-update-guide --file --sourceid 1 --xmlfile /tmp/xmltv.xml > /dev/null

RESULT=$?

if [ "$RESULT" -ne 0 ]; then
	echo "ERROR: mythfilldatabase failed with error code: $RESULT"
	exit -1
fi


tools/mheg5grab.sh
#!/bin/bash
# MHEG2XMLTV - mheg5grab.sh
# Wrapper script to run redbutton-download 
#
# Copyright (C) 2010-2016 Solorvox <solorvox@epic.geek.nz>
# 
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
# 
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.

# Time to run in seconds, 45-120 seconds is recommended
TIME="60" 

# Path to redbutton download
RB_DOWNLOAD="/usr/local/bin/rb-download"

# Base directory for rb-download dump MHEG-5 data
MHEG5_DIR="$HOME/mheg"

# Service ID
SERVICE_ID="1200"

# Channels configuration file (channels.conf)
CHANNELS="$HOME/.tzap/channels.conf"

# Adapter number, leave blank for default adapter 0
# "-a 1" is adapter 1 (or second tuner)
ADAPTER="-a 1"

# Full command line
COMMAND="$RB_DOWNLOAD $ADAPTER -f $CHANNELS -b $MHEG5_DIR $SERVICE_ID" 

# set channel for tuner before running (leave empty for none IE:  TUNEFIRST="")
# this will set the tuner to the correct frequency before running rbdownload.
TUNEFIRST="/usr/bin/ivtv-tune --freqtable=newzealand --channel=1 --device=/dev/video1"

# Remove old data before running, "TRUE" to remove $MHEG5_DIR/*
REMOVE_OLD="TRUE"
########### END CONFIG #############
# Check for redbutton binary
if [ \! -x $RB_DOWNLOAD ]; then
	echo "ERROR: Redbutton download not found at configuration setting [$RB_DOWNLOAD]." > /dev/stderr
	exit -1
fi

if [ "$TUNEFIRST" != "" ]; then
	$TUNEFIRST > /dev/null
fi

# Make the target directory if missing
if [ \! -d $MHEG5_DIR ]; then
	mkdir -p $MHEG5_DIR;
	if [ $? -ne 0 ]; then
		echo "ERROR: Couldn't find or create data directory [$MHEG5_DIR]." > /dev/stderr
		exit -1
	fi
fi

# Remove old data
if [ "$REMOVE_OLD" == "TRUE" ]; then
	rm -r $MHEG5_DIR/*
fi
	
# Run rb-download
${COMMAND} > /dev/null 2>&1 &
CPID=$!
RESULT="$?"

# If the pid is less than 1, something went wrong, exit
if [ "$CPID" -lt 1 ]; then
	echo "ERROR: Failed to run rb-download, command returned code [$RESULT]" > /dev/stderr
	exit -1
fi
	
# Delay in seconds
sleep $TIME

# Shutdown without terminated message
{ killall $( basename $RB_DOWNLOAD ); wait $CPID; } 2> /dev/null

exit 0


mheg2xmltv.sh
#!/bin/bash
# MHEG2XMLTV - Freeview DVB-T MHEG to XMLTV EPG converter
# Version: 0.5.0  20171130
# Copyright (C) 2010-2017  Solorvox <solorvox@epic.geek.nz>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.
#
# Requirements: rb-download, sed, and gawk

# Base directory for rb-download dump
SOURCE_DIR="$HOME/mheg"

# Output filename
OUT_FILE="/tmp/xmltv.xml"

# Service ID (1200 for NZ Freeview)
SERVICE_ID="1200"

# XMLTV Channel ID Replacement
# Format 'channel name,XML channel id,icon path
# Note: Channel name does not get modified, only channel id is changed.
# Modified so I don't have to change Myth settings
DEFAULT_CHANNEL_MAP='TV ONE,tv1.freeviewnz.tv,Channel-TV-One.png
TVONEPLUS1,tv1plus1.freeviewnz.tv,one-plus-one-h.png
TV2,tv2.freeviewnz.tv,Channel-Home-TV-2.png
TV33,tv33.freeviewnz.tv,Channel-Home-33TV.png
TV3,tv3.freeviewnz.tv,Channel-Home-TV-3.png
TV3PLUS1,tv3plus1.freeviewnz.tv,Channel-Home-TV-3Plus1.png
C4,c4.freeviewnz.tv,Channel-Home-C4.png
C42,c42.freeviewnz.tv,Channel-Home-C4.png
TV9,tv9.freeviewnz.tv,channel9.png
TV3Plus1,tv3-plus1.freeviewnz.tv,Channel-Home-TV-3Plus1.png
Māori Television,maori-tv.freeviewnz.tv,Channel-Home-Maori-TV.png
Parliament TV,parliament.freeviewnz.tv,Channel-Home-parliament-TV.png
PRIME,prime.freeviewnz.tv,Channel-Home-Prime.png
ChineseTV8,ctv8.freeviewnz.tv,Channel-Home-Chinese-TV-8.png
tvCentral,tvcentral.freeviewnz.tv,Channel-Home-TV-Central.png
ChoiceTV,topshelf.freeviewnz.tv,choice-home.png
TV Rotorua,rotorua.freeviewnz.tv,tv-rotorua.png
Info-Rotorua,info-rotorua.freeviewnz.tv,Channel-Home-info-Rotorua.png
Channel North Television,channelnorth.freeviewnz.tv,North.png
Trackside,trackside.freeviewnz.tv,Channel-Home-Trackside.png
FOUR,four.freeviewnz.tv,Channel-Home-Four.png
U,u.freeviewnz.tv,Channel-Home-U.png
RadioNZ National,rnz-national.freeviewnz.tv,Channel-Home-Radio-NZ-National.png
RadioNZ Concert,rnz-concert.freeviewnz.tv,Channel-Home-Radio-NZ-Concert.png
George,george-fm.freeviewnz.tv,Channel-Home-GeorgeFM.png
BaseFM,base-fm.freeviewnz.tv,ChannelsHome/basefm.png'

# Base URL for channel icons
CHANNEL_ICON_BASE="http://www.freeviewnz.tv/media/nonumbracoimages/original/ChannelsHome/"

# Metadata file, used for optional information such as series categories
# Default same directory as script.
METADATA="$(pwd)/series-metadata.txt"

# Temp file used during conversion (default mktemp)
TMP_STAGE1=`mktemp`
TMP_STAGE2=`mktemp`
TMP_STAGE3=`mktemp`

# Clean up older data (use with caution)
# Default CLEANUP="find $SOURCE_DIR -ctime +8 -delete"
#CLEANUP="find $SOURCE_DIR -ctime +8 -delete"

## END CONFIG ##

USAGE="$0 [-o output_file] [-i input_directory] [-s service_id] [-h] [-d] [-x] [-v]
	  [-m metadata file] 
	-h	Help/usage
	-v	Be verbose
	-o	Output filename (default /tmp/xmltv.xml)
	-i	Input data directory (default \$HOME/mheg)
	-s	Service ID (default 1200)
	-x	Use crid://name instead of service id (eg 1200) for channel ID
	-d	Use default channel remapping (default none)
	-m	Metadata file, additonal infomation such as catagories.
	-c	Channel map file, overides default values.
	-g	Guess at show data from title/description (Starring = Movie/etc)"

VERBOSE=0
CHANNEL_ID_FIELD="ServiceID"
# Parse cmd line
while getopts ":c:m:o:i:s:hxdvg" opt
do
	case $opt in
 	v )
		VERBOSE=1
		;;	
	c )
		CHANNEL_MAP_FILE="$OPTARG"
		;;
	m )
		METADATA="$OPTARG"
		;;
	o )
		OUT_FILE="$OPTARG"
		;;
	i )
		SOURCE_DIR="$OPTARG"
		;;
	s )
		SERVICE_ID="$OPTARG"
		;;
	h )
		echo "Usage: $USAGE" >&2
		exit 0
		;;
	x )
		CHANNEL_ID_FIELD="crid"
		;;
	g )
		GUESS="YES"
		;;
	d )
		CHANNEL_MAP=$DEFAULT_CHANNEL_MAP
		;;
	? )
		echo "Invalid option: -$OPTARG" >&2
		echo "Usage: $USAGE" >&2
		exit -1
		;;
	esac
done

if [ "$GUESS" == "YES" -a \! -e $METADATA ]; then
	echo "ERROR: Metatdata file is not found.  Either supply one via command line or ensure it is in the current path.";
	exit -1
fi

if [ "$TMP_STAGE1" == "" -o "$TMP_STAGE2" == "" -o "$TMP_STAGE3" == "" ]; then
	echo "ERROR: mktemp failed to allocate temp files ([$TMP_STAGE1], [$TMP_STAGE2], or [$TMP_STAGE3]), aborting."
	exit -1
fi

# Default BASE="$SOURCE_DIR/services/$SERVICE_ID/epg/data"
BASE="$SOURCE_DIR/services/$SERVICE_ID/epgdtt/data"

if [ \! -d "$BASE" ]; then
	echo "ERROR: Source directory [$SOURCE_DIR] does not contain expected data, aborting conversion."
	exit -1
fi

if [ "$CHANNEL_MAP" != "" -o "$CHANNEL_MAP_FILE" != "" ]; then
	if [ "$VERBOSE" -gt 0 ]; then
		echo "Remapping channels using channel map";
	fi
	if [ "$CHANNEL_MAP_FILE" != "" ]; then
		CHANNEL_MAP="$( cat $CHANNEL_MAP_FILE )"	
	fi
	# Build sed channel remap command
	OLDIFS=$IFS
	IFS=$(echo -en "\n\b")
	SEDREMAP="sed "
	for cidmap in ${CHANNEL_MAP}; do
		chan_name=$( echo $cidmap | awk -F, '{printf "%s\n", $1}' )
		new=$( echo $cidmap | awk -F, '{printf "%s\n", $2}' )
		icon=$( echo $cidmap | awk -F, '{printf "%s\n", $3}' )
		SEDREMAP="$SEDREMAP -e "\'"/$chan_name.crid/s|\(crid:[^\x1d]*\)|$new\x1d\1\x1d$icon|"\'""
	done
	IFS=$OLDIFS
fi

# Loop through all the days of data
for day in $( ls "$BASE" );
do
	# Loop for all channels
	for chan in $( ls $BASE/$day/* )
	do
		sid=`basename $chan`
		# Dump everything into master file, remapping in process
		if [ "$CHANNEL_MAP" != "" ]; then
			eval "$SEDREMAP $chan" >> $TMP_STAGE1
		else
			# replace crid with serviceID for easy tagging in myth
			sed "/crid:/s/\x1c/$sid\x1d\x1c/" $chan >> $TMP_STAGE1
		fi
		echo >> $TMP_STAGE1 # extra CR for file break
	done
done

# Reformat MHEG data
# 1c = end of record
# 1d = field seperator
sed -i 's/[\x0a\x0d]/ /g;s/\x1c/\n/g;s/[\x04\x00]//g;s/\x1b[Cc]//g;s/&/&amp;/g' $TMP_STAGE1

# DEBUG
cp $TMP_STAGE1 /tmp/stage1

if [ $? -ne 0 ]; then
	echo "ERROR: Failed reformat data file, aborting."
	exit -1
fi

# Create header and channel list from all channels found in MHEG stream
gawk -f- $TMP_STAGE1 > $OUT_FILE <<EOF
BEGIN {
	FS="\x1d"
	print "<tv date=\"" strftime("%Y%m%d%H%M%S %z", systime()) "\" generator-info-name=\"mheg2xmltv\" "
	print "source-info-name=\"DVB-T MHEG Stream\">"

	channel_ids[""]=0
	channel_count=0
}
{
	if ( \$0 ~ /crid:/)
	{
		displayName=\$3
		i=0;
		cid=\$6
			
		for (chan in channel_ids)
		{
			if (chan == cid )
				break
			else
				i++;
		}

		if (i >= channel_count) 
		{
			channel_ids[cid]=i
			channel_count++

			print "\t<channel id=\"" cid "\">"
			print "\t\t<display-name>" displayName "</display-name>"

			if ( \$7 != "" ) 
				print "\t\t<icon src=\"$CHANNEL_ICON_BASE" \$7 "\" />"

			print "\t</channel>"
		}
	}
}
EOF

if [ "$VERBOSE" -gt 0 ]; then
	
	echo -n "Channels: "
	grep "display-name" $OUT_FILE | sed 's/.*<display-name>\(.*\)<.*$/\1/' | tr '\n' ','
	echo 
fi
# Main event listings, written to temporary stage2 file for further processing
gawk -f- $TMP_STAGE1 > $TMP_STAGE2 <<EOF
BEGIN {

	metadata_categories[""]=0
	metadata_showid_to_title[""]=0

	# Read in the metadata on shows if enabled
	if ( "$METADATA" != "" )
	{
		FS="|"
		while (( getline line < "$METADATA") > 0)
		{
			# Ignore comments, read in metadata
			if (line !~ /^#/)
			{
				split(line, data, "|")
				metadata_categories[data[1]]=data[4]

				# now store the showid into an assoc array mapped to show title
				split(data[3], ids, ",")
				for (i in ids)
					metadata_showid_to_title[ ids[ i ] ]=data[1]
			}
		}
	}

	FS="\x1d"
}
{
	if ( ( \$4 ~ /crid:/ ) || ( \$5 ~ /crid:/ ) )
	{
		chan=\$4
		cid=\$6

		date_conv="date +\"%Y %m %d\" -d \"" \$2 "\""
		date_conv | getline file_date
		close(date_conv)
		date=file_date " 0 0 0"
		next
	}

	# if data is blank, just skip this record
	if ( \$8 == "" ) { next }

	start_time=strftime("%Y%m%d%H%M00 %z", mktime(date) + \$2)
	end_time=strftime("%Y%m%d%H%M00 %z", mktime(date) + \$3)
	sub(/:/, "", \$5)
	show_record=start_time "\t" end_time "\t" cid "\t" \$8 "\t" \$9

	# Episode ID
	if ( \$7 != "" )
	{
		ep_num = \$7
		gsub(/^\//,"", ep_num)
		show_record=show_record "\t" ep_num
	}
	else
	{
		show_record=show_record "\t<blank>"
	}

	# find the show ID number, starting after field 9
	for (i=9; i < NF; i++)
	{
		if ( (\$i ~ /^\//) && (\$i !~ /\.png/) )
			break;
	}
	if (i < NF)
	{
		showid=\$i
		gsub(/^\//,"", showid)
	}

	icon_count=\$10

	# Match icons with known tags
	if ( \$0 ~ /hd\.png/ )
		show_record=show_record "\tyes_hd"
	else
		show_record=show_record "\tno_hd"

	if ( \$0 ~ /dolby\.png/ )
		show_record=show_record "\tyes_dolby"
	else
		show_record=show_record "\tno_dolby"

	if ( \$0 ~ /ear\.png/ )
		show_record=show_record "\tyes_ear"
	else
		show_record=show_record "\tno_ear"

	if ( \$0 ~ /ao\.png/ )
		show_record=show_record "\tao"
	else if ( \$0 ~ /g\.png/ )
		show_record=show_record "\tg"
	else if ( \$0 ~ /pgr\.png/ )
		show_record=show_record "\tpgr"
	else
		show_record=show_record "\tno_rating" 

	# Category fields, try to match on showid first (best match)
	catcount=0
	if ( showid in metadata_showid_to_title )
	{
		show_title=metadata_showid_to_title[ showid ]
		if ( show_title != "")
		{
			split(metadata_categories[ show_title ], categories, ",")
			for (cat in categories)
				show_record=show_record "\t" categories[cat]
		}
	}
	else
 	{
		# fall-back if not matched by showid (second best)
		if ( \$8 in metadata_categories )
		{
			split(metadata_categories[\$8], categories, ",")
			for (cat in categories)
				show_record=show_record "\t" categories[cat]
		}
		else
		{
			# if all else fails, and enabled by command line, guess at it. :)
			if ( "$GUESS" == "YES" )
			{
				title=\$8
				description=\$9
				
				# Generic series
				if (( description ~ /[Ee]pisode/ ) || ( description ~ /[Ss0-9] ?[Ee0-9]/ ))
					show_record=show_record "\tSeries"
				
				# News
				if ( title ~ /[Nn]ews/ )
					show_record=show_record "\tNews"

				# Shopping
				if (( title ~ /[Ss]hopping/ ) || ( title ~ /[Ii]nfomercial/ ))
					show_record=show_record "\tShopping"

				# Spiritual
				if (( title ~ /[Rr]eligion/ ) || ( title ~ /[Rr]eligion/ ) \
					|| ( description ~ /[Rr]eligious [Pp]rogram/) )
					show_record=show_record "\tSpiritual"

				# Travel
				if ( title ~ /[Tt]ravel/ )
					show_record=show_record "\tTravel"
					
				# Movies
				if (( description ~ /Starring/ ) || ( description ~ /[Ss]tars /) \
					|| ( title ~ /Movie:/ ) )
					show_record=show_record "\tMovie"

				# Documentary
				if (( title ~ /[Dd]ocumentary/ ) || ( description ~ /[Dd]ocumentary/ ))	
					show_record=show_record "\t Documentary"

				# Sports
				if (( title ~ /Sport/ ) || ( title ~ /playoff/ ) || ( description ~ /[Ss]ports news/) \
					|| ( title ~ /[Rr]ugby/ ) || ( title ~ /NFL|NBA|NHL[ -]/ ))
					show_record=show_record "\tSports"

				# Comedy
				if ( description ~ /[Cc]omedy/ )
					show_record=show_record "\tComedy"

				# Drama
				if (( description ~ /[Dd]rama/ ) || ( title ~ /[Dd]rama:/ ))
					show_record=show_record "\tDrama"

				# Children
				if ( description ~ /[Cc]hildren [Ss]how/ )
					show_record=show_record "\tChildren"

				# SciFi
				if (( title ~ /[Ss]ci-?[Ff]i/ ) || ( title ~ /[Ss]ci-?[Ff]i/ ))	
					show_record=show_record "\tSciFi"

				# cooking
				if ( ( title ~ /[Cc]ooking/ ) || ( title ~ /[Cc]heif/ ) )
					show_record=show_record "\tFood"

				# Business
				if ( title ~ /[Bb]usiness/ )
					show_record=show_record "\tBusiness"
				
				# Game Show
				if (( title ~ /[Gg]ame [Ss]how/) || ( description ~ /[Gg]ame [Ss]how/) \
					|| ( description ~ /contestant/ ) )
					show_record=show_record "\tGame"

				# Reality
				if ( description ~ /reality series/ )
					show_record=show_record "\tReality\tSeries"

				# Nature/Science
				if (( title ~ /[Ss]cicence/ ) || ( title ~ /[Nn]ature/ ) )
					show_record=show_record "\tScience"


				# Talk Show
				if (( description ~ /daytime talk/ ) || ( description ~ /talk show/ ))
					show_record=show_record "\tTalk"
				
				# Crime 
				if (( description ~ /criminal investigation/ ) \
					|| ( description ~/case.*murder/ ) )
					show_record=show_record "\tCrime"
				
			} 
		}
	}

	print show_record
}
EOF

# Fix for multi-day spanning shows, stage2 -> stage3
gawk -f- $TMP_STAGE2 > $TMP_STAGE3 <<EOF
BEGIN { FS = "\t" } # Use tab file separator
{
	# Save fields we'll need later
	start_time=\$1
	end_time=\$2
	cid=\$3
	title=\$4
	episode_id=\$6
	wholerec=\$0

	# Array for duplicate entries to delete
	delete_list[""]=0

	#  Does the show supposedly finish at midnight?
	if ( end_time ~ /000000 / ) # Note space for ending of time
	{
		found_end=0
		# Loop through Stage2 again. Note field numbers now apply to this new file, not parent file read loop.
		while ((getline < "$TMP_STAGE2") >0)
		{
			# Is there another show starting at the end_time from the first read
			# and on the same channel?
			if (( \$1 == end_time ) && ( \$3 == cid ))
			{
				# Do the title and episode_id match?
				if (( \$4 == title ) && ( \$6 == episode_id ))
				{
					# Read the correct end time from the second listing
					correct_end=\$2
					found_end=1
				}
				else
				{
					# If and only if we find a different show starting at midnight then we
					# can write this show ending at midnight
					print wholerec
				}
				break
			}
		}
		close("$TMP_STAGE2")

		# if we found and fixed an end time
		if ( found_end == 1)
		{
			# Strip off the start/end times and print corrected end time with remainder of record.
			pos=length(start_time) + length(correct_end) + 2;
			trailer=substr(wholerec, pos + 1, length(wholerec)- pos)
			print start_time "\t" correct_end "\t" trailer

			# add entry to the delete list so duplicate will be removed
			delete_list[ delete_count++ ] = title "|" episode_id "|" correct_end
		}
	}

	# Does the show start at midnight?
	else if (\$1 ~ /000000 / ) # Note space for match on end of time string
	{
		# Only print records starting at midnight if the preceding record has a different title 
		found_diff=0

		for (i = 0; i <= delete_count; i++)
		{
			key=title "|" episode_id "|" end_time	
			if ( delete_list[ i ] == key )
				found_diff=1
		}

		# if a match wasn't found then print the record
		if ( found_diff == 0)
			print \$0
	}
	else
	{
		# Record didn't start at midnight, print
		print \$0
	}
}
EOF

# Final main event listings, appended to file because of channel header
# Expected input format (tab seperated)
#      1         2        3      4        5          6         7        8          9           10      11      ...
# start_time stop_time channel title description episode_id hd_flag dolby_flag teletext_flag rating categories ...
gawk -f- $TMP_STAGE3 >> $OUT_FILE <<EOF
BEGIN { FS="\t" }
{
	print "\t<programme channel=\"" \$3 "\" start=\"" \$1 "\" stop=\"" \$2 "\">"
	print "\t\t<title>" \$4 "</title>"
	print "\t\t<desc>" \$5 "</desc>"

	# Episode ID
	if ( \$6 != "<blank>" )
	{
		print "\t\t<episode-num system=\"dd_progid\">" \$6 "</episode-num>"
	}

	for (i=11; i<=NF; i++)
	{
		if ( \$i != "<blank>" )
			print "\t\t<category>" \$i "</category>"
	}

	# Match icons with known tags
	if ( \$7 == "yes_hd" )
	{
		print "\t\t<video>"
		print "\t\t\t<present>yes</present>"
		print "\t\t\t<quality>HDTV</quality>"
		print "\t\t</video>"
	}

	if ( \$8 == "yes_dolby" )
	{
		print "\t\t<audio>"
		print "\t\t\t<stereo>dolby</stereo>"
		print "\t\t</audio>"
	}

	if ( \$9 == "yes_ear" )
	{
		print "\t\t<subtitles type=\"teletext\" />"
	}

	if ( \$10 == "ao" )
	{
		print "\t\t<rating system=\"Freeview\">"
		print "\t\t\t<value>AO</value>"
		print "\t\t</rating>"
	}
	else if ( \$10 == "g" )
	{
		print "\t\t<rating system=\"Freeview\">"
		print "\t\t\t<value>G</value>"
		print "\t\t</rating>"
	}
	else if ( \$10 == "pgr" )
	{
		print "\t\t<rating system=\"Freeview\">"
		print "\t\t\t<value>PGR</value>"
		print "\t\t</rating>"
	}
	print "\t</programme>"
}
END { print "</tv>" }
EOF

# Clean-up temp files and then run cleanup command
rm $TMP_STAGE1 $TMP_STAGE2 $TMP_STAGE3
$CLEANUP
